From 8b92f598720a4a352d2ff66bc27fe69d34df7a45 Mon Sep 17 00:00:00 2001
From: allen yan <yanwei@marvell.com>
Date: Thu, 16 Feb 2017 01:10:15 +0800
Subject: [PATCH 0789/1345] uart: a3700: remove unrelated reg bits setting for
 baudrate divisor

commit  28154891ebccb50cfdc5b5b55644d14b0c7f8db6 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- In baudrate divisor setting, not only the relevant bits 0 - 9
  of register 0xd0012010/0xd0012020 are set, but also all the rest
  bits are incorrectly cleared.
- This patch fixes this issue and removes the irrelevant bits setting.

Change-Id: I6dfe430e242fe4b57a7dee95f04bae63d591fcee
Signed-off-by: allen yan <yanwei@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/36812
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hua Jing <jinghua@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/tty/serial/mvebu-uart.c |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/mvebu-uart.c b/drivers/tty/serial/mvebu-uart.c
index e5f934f..e3cb718 100644
--- a/drivers/tty/serial/mvebu-uart.c
+++ b/drivers/tty/serial/mvebu-uart.c
@@ -101,6 +101,8 @@ struct uart_regs_layout {
 				 | STAT_PAR_ERR | STAT_OVR_ERR)
 
 #define UART_BRDV		0x10
+#define  BAUD_MASK		0x000003ff
+#define  BAUD_OFFSET		0
 
 /* REG_UART_A3700_EXT */
 #define UART_EXT_CTRL		0x04
@@ -554,6 +556,7 @@ static void mvebu_uart_shutdown(struct uart_port *port)
 
 static void mvebu_uart_baud_rate_set(struct uart_port *port, unsigned int baud)
 {
+	unsigned int value;
 	unsigned int baud_rate_div;
 	struct mvebu_uart_data *uart_data = (struct mvebu_uart_data *)port->private_data;
 
@@ -577,7 +580,10 @@ static void mvebu_uart_baud_rate_set(struct uart_port *port, unsigned int baud)
 	 */
 	if (!IS_ERR(uart_data->clk)) {
 		baud_rate_div = DIV_ROUND_UP(port->uartclk, (16 * baud));
-		writel(baud_rate_div, port->membase + REG_BRDV(uart_data));
+		value = readl(port->membase + REG_BRDV(uart_data));
+		value &= ~BAUD_MASK;
+		value |= baud_rate_div << BAUD_OFFSET;
+		writel(value, port->membase + REG_BRDV(uart_data));
 	}
 }
 
-- 
1.7.9.5

